### PR TITLE
Add Modal Interactions and Input Text

### PR DESCRIPTION
**Pull request by pascalharp**
Modals are out of beta

This PR adds modal interactions and input text options.

# Docs
Modals: https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-response-object-modal
Text Input: https://discord.com/developers/docs/interactions/message-components#text-inputs

### TODO

- [x]  Add modal interactions
- [x]  Add input text
- [x]  Add input text builder
- [x]  Finish ModalSubmitInteractions
- [x]  Testing: https://github.com/pascalharp/serenity_modal_testing
- [x]  Fix merge conflicts

### CODE REVIEW

**Comment by ghost** [comment]
it seems that `ModalSubmit` has a slightly different structure to normal `MessageComponent`s, containing a `components` array ([see here](https://github.com/discord/discord-api-docs/commit/4b6c6a14d46ff2e62f6384e3ebd9116243c650c9#diff-b44c2bafd9f1fb4cd3383e25770ef07d972796aa56dedd73ba1a7853df2d62d0R60)), and that's where you get what they responded with

**Comment by pascalharp** [comment]
Some initial testing: https://github.com/pascalharp/serenity_modal_testing

**Comment by pascalharp** [comment]
It's working so far but I am a bit unsure about the structure since Modal- and Component Interactions blur the line quite a bit. They are different interaction types and the fields differ as well. However both share `ActionRows` and `InputText` is considered a Component just like 'Button' and `SelectMenu` according to the docs. Even if they can never exist together.

Right now `ModalSubmitInteraction.data.components` uses the `ActionRow` implementation in `message_component.rs.`
It might be better to fully separate and have some code duplication or maybe move ActionRow and Components into a separate file.

Happy for any input and suggestion =D

**Comment by Milo123459** [comment]
I think using the implementation from the ActionRow is fine as opposed to duplicating the code.

File: src/model/interactions/message_component.rs [hunk 1]
```diff
@@ -675,3 +695,31 @@ pub struct SelectMenuOption {
     #[serde(default)]
     pub default: bool,
 }
+
+/// A input text component for modal interactions
```
Comment (Flat) [inline src/model/interactions/message_component.rs]:
```suggestion
/// An input text component for modal interactions
```

File: src/model/interactions/message_component.rs [hunk 2]
```diff
@@ -675,3 +695,31 @@ pub struct SelectMenuOption {
...
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct InputText {
+    /// The component type, it will always be [`ComponentType::InputText`].
+    #[serde(rename = "type")]
+    pub kind: ComponentType,
+    /// An identifier defined by the developer for the select menu.
+    pub custom_id: String,
+    /// The Input from the user
```
Comment (Flat) [inline src/model/interactions/message_component.rs]:
```suggestion
    /// The input from the user
```

**Review by Flat — APPROVED** [review]
Minor grammar nits. Looks good to me, have not tested the code.

File: src/model/event.rs [hunk 3]
```diff
@@ -2239,24 +2239,30 @@ macro_rules! with_related_ids_for_event_types {
...
+                    Interaction::ModalSubmit(i) => Some(i.channel_id),
                 },
                 message_id: match &e.interaction {
                     Interaction::Ping(_) => None,
                     Interaction::ApplicationCommand(_) => None,
                     Interaction::MessageComponent(i) => Some(i.message.id),
                     Interaction::Autocomplete(i) => None,
+                    Interaction::ModalSubmit(i) => {
+                        if let Option::Some(m) = &i.message { Some(m.id) } else { None }
+                    },
```
Comment (arqunis) [inline src/model/event.rs]:
```suggestion
                    Interaction::ModalSubmit(i) => i.message.as_ref().map(|m| m.id),
```

**Review by arqunis — CHANGES_REQUESTED** [review]

File: src/model/event.rs (hunk 3 above)
Comment (pascalharp) [reply src/model/event.rs]: map doesn't seem to work here. Since it expects a RelatedId and not Option.

Comment (arqunis) [reply src/model/event.rs]: Ah, I forgot `RelatedId` borrowed `Option`'s variants; however, you can still use the `map` by converting the `Option` to `RelatedId` (`RelatedId` has a `From<Option<T>>` impl):
```suggestion
                    Interaction::ModalSubmit(i) => i.message.as_ref().map(|m| m.id).into(),
```

**Review by arqunis — APPROVED** [review]

**Pull request merged**